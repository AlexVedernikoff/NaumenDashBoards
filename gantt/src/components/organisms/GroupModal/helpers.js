// @flow
import type {AndCondition, OrCondition, SubGroup} from 'GroupModal/types';
import type {Attribute} from 'store/sources/attributes/types';

/**
 * Создает новое условие ИЛИ
 * @returns {OrCondition}
 */
const createNewOrCondition = (): OrCondition => ({
	type: ''
});

/**
 * Создает новое условие И
 * @returns {AndCondition}
 */
const createNewAndCondition = (): AndCondition => [createNewOrCondition()];

/**
 * Создает новую подгруппу
 * @returns {SubGroup}
 */
const createNewSubGroup = (): SubGroup => ({
	data: [createNewAndCondition()],
	name: ''
});

/**
 * Создает тип пользовательской группировки для ссылочного атрибута
 * @param {Attribute} attribute - атрибут
 * @param {string} property - свойства которое нужно использовать, для создания уникальности
 * @returns {string}
 */
const createRefObjectCustomGroupType = (attribute: Attribute, property: 'metaClassFqn' | 'property'): string => {
	return `${attribute.type}$${attribute[property]}`;
};

export {
	createRefObjectCustomGroupType,
	createNewAndCondition,
	createNewOrCondition,
	createNewSubGroup
};
