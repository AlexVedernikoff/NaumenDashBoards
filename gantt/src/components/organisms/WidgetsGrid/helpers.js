// @flow
import {DEFAULT_COLS_COUNT, GRID_PROPS, gridRef} from './constants';
import {DEFAULT_WIDGET_LAYOUT_SIZE} from 'store/dashboard/layouts/constants';
import type {Layout, WidgetLayoutPosition} from 'store/dashboard/layouts/types';
import type {LayoutBreakpointsData, LayoutMode} from 'store/dashboard/settings/types';

/**
 * Возвращает количество колонок сетки в соотвествии с текущим модом дашборда и шириной
 * @param {LayoutMode} layoutMode - мод отображения дашборда
 * @returns {number} - количество колонок сетки
 */
const getCountGridColumns = (layoutMode: LayoutMode) => {
	const {breakpoints, cols} = GRID_PROPS[layoutMode];
	const {current: grid} = gridRef;
	let count = DEFAULT_COLS_COUNT;

	if (grid) {
		const {clientWidth: width} = grid;
		const sortedKeys = Object.keys(breakpoints).sort((a, b) => (breakpoints[b] - breakpoints[a]));
		let breakpoint = sortedKeys[0];

		sortedKeys.forEach(key => {
			if (breakpoints[key] > width) {
				breakpoint = key;
			}
		});

		if (cols[breakpoint]) {
			count = cols[breakpoint];
		}
	}

	return count;
};

/**
 * Формирует узкий макет из широкого
 * @param {Layout[]} layouts - широкий макет
 * @param {Layout[][]} oldLayouts - старый узкий макет
 * @returns  {Layout[]} - новый узкий макет
 */
const generateWebSMLayout = (layouts: Layout[], oldLayouts: ?Layout[]): Layout[] => {
	if (layouts && layouts.length > 0) {
		const heights = {};
		const result = [];

		layouts.forEach(({h, i}) => {
			heights[i] = h;
		});
		oldLayouts && oldLayouts.forEach(({h, i}) => {
			heights[i] = h;
		});

		let top = 0;
		const widgetIdsByPosition = layouts.map(({i, x, y}) => ({i, pos: y * 100 + x}));

		widgetIdsByPosition.sort(({pos: first}, {pos: second}) => first - second);
		widgetIdsByPosition.forEach(({i}) => {
			const h = (i in heights) ? heights[i] : DEFAULT_WIDGET_LAYOUT_SIZE.h;

			result.push({ h, i, w: 1, x: 0, y: top });
			top += h;
		});

		return result;
	}

	return oldLayouts ?? [];
};

/**
 * Сравнивает 2 макета
 * @param {Layout[]} firstLayouts - первый макет
 * @param {Layout[]} secondLayouts - второй макет
 * @returns  {boolean}
 */
const isEqualsLayouts = (firstLayouts: ?Layout[], secondLayouts: ?Layout[]): boolean => {
	if (!firstLayouts || !secondLayouts || firstLayouts.length !== secondLayouts.length) {
		return false;
	}

	return firstLayouts.every((firstLayout) => {
		const secondLayout = secondLayouts.find(({i}) => i === firstLayout.i);
		return secondLayout
			&& firstLayout.h === secondLayout.h
			&& firstLayout.w === secondLayout.w
			&& firstLayout.x === secondLayout.x
			&& firstLayout.y === secondLayout.y;
	});
};

/**
 * Расчитывает позицию клика по параметрам сетки
 * @param {LayoutMode} layoutMode - режим отображения дашборда (веб/мобильный)
 * @param {LayoutBreakpointsData<number>} breakpoints - список режимов сетки по ширине элемента
 * @param {LayoutBreakpointsData<number>} colsCount - количество колонок в зависисмости от режима сетки
 * @param {number} rowHeight - высота строки сетки
 * @param {number} width - ширина элемента сетки
 * @param {number} mouseX - X координата клика
 * @param {number} mouseY - Y координата клика
 * @returns {WidgetLayoutPosition} - расчетное положение клика
 */
const calculatePosition = (
	layoutMode: LayoutMode,
	breakpoints: LayoutBreakpointsData<number>,
	colsCount: LayoutBreakpointsData<number>,
	rowHeight: number,
	width: number,
	mouseX: number,
	mouseY: number
): ?WidgetLayoutPosition => {
	const keys = Object.keys(breakpoints).sort((bp1, bp2) => breakpoints[bp1] - breakpoints[bp2]);
	const breakpoint = keys.find((breakpoint, idx) => {
		const nextBreakpoint = keys[idx + 1];
		return (nextBreakpoint === undefined
			|| (breakpoints[breakpoint] <= width && width <= breakpoints[nextBreakpoint]));
	}) ?? keys[0];

	if (breakpoint in colsCount) {
		const colWidth = width / colsCount[breakpoint];
		const x = Math.trunc(mouseX / colWidth);
		const y = Math.trunc(mouseY / rowHeight);

		return {breakpoint, layoutMode, x, y};
	}

	return null;
};

export {
	getCountGridColumns,
	generateWebSMLayout,
	isEqualsLayouts,
	calculatePosition
};
