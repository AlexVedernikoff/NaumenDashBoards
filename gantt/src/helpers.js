// @flow
/**
 * Возвращает функцию, которая не будет срабатывать, пока продолжает вызываться
 * @param {Function} func - функция которую необходимо обернуть
 * @param {number} ms - количество миллисекунд для проверки повторного вызова
 * @returns {Function} - обернутая функция
 */
function debounce (func: Function, ms: number) {
	let timer = null;

	return function () {
		clearTimeout(timer);
		timer = setTimeout(() => func(...arguments), ms);
	};
}

/**
 * Функция для обхода ошибки flow о несоответствии типов при использовании Object.values
 * @param {object} map - объект
 * @returns {Array<any>} - массив значений объекта
 */
function getMapValues<T> (map: $Shape<{[key: string]: T}>): Array<T> {
	return Object.keys(map).map(key => map[key]);
}

/**
 * Функция проверяет является ли переменная объектом
 * @param {any} item - проверяемая переменная
 * @returns {boolean}
 */
const isObject = (item: any): boolean => item && typeof item === 'object' && !Array.isArray(item);

/**
 * Объединяет 2 объекта с учетом вложенности
 * @param {object} target - исходный объект
 * @param {object} source - объект, содержащий значения, которые необходимо заменить в исходном объекте
 * @returns {object}
 */
const extend = (target: Object, source: Object): Object => {
	let output = {...target};

	if (isObject(target) && isObject(source)) {
		Object.keys(source).forEach(key => {
			if (isObject(source[key])) {
				if (key in target) {
					output[key] = extend(target[key], source[key]);
				} else {
					output = {...output, [key]: source[key]};
				}
			} else {
				output = {...output, [key]: source[key]};
			}
		});
	}

	return output;
};

/**
 * Реализует глубокое копирование объекта
 * @param {object} object - исходный объект
 * @returns {object}
 */
const deepClone = (object: Object) => JSON.parse(JSON.stringify(object));

/**
 * Реализует конвертацию даты для обхода FireFox
 * @param {string} date - исходная дата
 * @returns {object}
 */
const normalizeDate = date => {
	const [datePart, time] = date.split(',');
	const dateParts = datePart.replace(/[\.\-\/]/g, ',').split(',');

	if (dateParts[2].length === 4) {
		[dateParts[0], dateParts[2]] = [dateParts[2], dateParts[0]];
	}

	const modifiedDateString = time ? dateParts.join('/') + time : dateParts.join('/');
	const modifiedDate = new Date(modifiedDateString);

	return modifiedDate;
};

/**
 * Вырезает смещение из даты
 * @param {Date} date - дата с выставленным timeZone
 * @returns {string} - значение смещения
 */
const shiftTimeZone = (date: Date) => {
	if (date) {
		const timezone = /(GMT.*\))/.exec(new Date(date));
		const deviation = timezone[0].slice(5, 6);
		const sign = timezone[0].slice(3, 4);
		let deleteDeviation = `${deviation}`;

		if (sign === '-') {
			deleteDeviation = `'-'${deviation}`;
		}

		return deleteDeviation;
	}
};

/**
 * Проверяет относится ли операционная система пользователя к семейству MasOS
 * @returns {boolean}
 */
const isMacOS = () => ['MacIntel', 'Mac68K', 'MacPPC'].includes(navigator.platform);

/**
 * Проверяет объект на пустоту
 * @returns {boolean}
 */
const isEmpty = obj => Object.keys(obj).length === 0;

/**
 * На текущий момент у окружения метода нет полифилов к es6. Для стабильной работы на старых браузерах все ответы
 * перепарсиваются согласно окружению внутреннего приложения.
 * @param {Function} restCallModule - метод для осуществления запросов к модулям
 * @returns {void}
 */
const decorateRestCallModule = (restCallModule: Function) => {
	window.jsApi.restCallModule = async (...params) => {
		const data = await restCallModule(...params);

		return JSON.parse(JSON.stringify(data));
	};
};

/**
 * Экранирует специальные символы строки
 * TODO: RegExp.esacape https://github.com/tc39/proposal-regex-escaping
 * @param {string} string - строка
 * @returns {string}
 */
const escapeString = (string: string) => string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

/**
 * Создает копию объета obj без поля [key]
 * @param {object} obj - объект для копирования
 * @param {string} key - ключ объекта obj
 * @returns {object} - копия obj без ключа key
 */
function omit<T: Object> (obj: T, key: string): T {
	if (key in obj) {
		const newObj = {...obj};

		delete newObj[key];

		return newObj;
	}

	return obj;
}

export {
	debounce,
	decorateRestCallModule,
	deepClone,
	escapeString,
	extend,
	getMapValues,
	isEmpty,
	isMacOS,
	isObject,
	normalizeDate,
	omit,
	shiftTimeZone
};
