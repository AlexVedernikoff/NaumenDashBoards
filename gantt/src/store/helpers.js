// @flow
import {store} from 'app.constants';

const isSourceType = (classFqn: string) => classFqn.includes('$');

/**
 * Возвращает подтипы источника
 * @param {string} classFqn - код источника
 * @returns {Array<string>}
 */
const getSourceTypes = (classFqn: string) => {
	const {sources} = store.getState();
	let types = sources[classFqn].children || [];

	if (types.length > 0) {
		const subTypes = types.map(getSourceTypes)
			.reduce((allSubTypes, subTypes) => [...allSubTypes, ...subTypes], []);

		types = [...types, ...subTypes];
	}

	return types;
};

/**
 * Возвращает массив для окна фильтрации, содержащий код источника и все его подтипы
 * @param {string} classFqn - код источника
 * @returns {Array<string>}
 */
const getDescriptorCases = (classFqn: string) => [classFqn, ...getSourceTypes(classFqn)];

const createFilterContext = (classFqn: string) => {
	const context: Object = {};

	if (isSourceType(classFqn)) {
		context.cases = getDescriptorCases(classFqn);
	} else {
		context.clazz = classFqn;
	}

	return context;
};

const getFilterContext = (descriptor: string, classFqn: string) => {
	let context = JSON.parse(descriptor);

	if (!context.clazz) {
		context = {
			...context,
			cases: getDescriptorCases(classFqn)
		};
	}

	return context;
};

export {
	createFilterContext,
	getFilterContext
};
