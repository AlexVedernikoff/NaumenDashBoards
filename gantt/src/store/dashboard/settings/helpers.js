// @flow
import {LOCAL_STORAGE_VARS} from 'store/constants';

const initPersonalProperty = () => localStorage.getItem(LOCAL_STORAGE_VARS.PERSONAL_DASHBOARD) === 'true';

export {
	initPersonalProperty
};
