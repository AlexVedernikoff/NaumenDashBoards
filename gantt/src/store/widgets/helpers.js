// @flow
import type {AnyWidget, Group, MixedAttribute, Widget, WidgetType} from './data/types';
import type {Attribute} from 'store/sources/attributes/types';
import {ATTRIBUTE_SETS, ATTRIBUTE_TYPES} from 'store/sources/attributes/constants';
import {
	DATETIME_SYSTEM_GROUP,
	DEFAULT_AGGREGATION,
	DEFAULT_SYSTEM_GROUP,
	GROUP_WAYS,
	INTEGER_AGGREGATION,
	INTERVALS
} from './constants';
import {DISPLAY_MODE, WIDGET_TYPES} from './data/constants';
import type {LayoutMode} from 'store/dashboard/settings/types';
import {LAYOUT_MODE} from 'store/dashboard/settings/constants';
import NewWidget from 'store/widgets/data/NewWidget';

const createDefaultGroup = (data?: string | null, attribute?: Attribute) => {
	if (!data || typeof data !== 'string') {
		return getDefaultSystemGroup(attribute);
	}

	return ({
		data,
		way: GROUP_WAYS.SYSTEM
	});
};

const getDefaultSystemGroup = (attribute?: Object) => {
	const target = attribute?.ref || attribute;

	if (target) {
		const {type} = target;

		if (type === ATTRIBUTE_TYPES.dtInterval) {
			return createDefaultGroup(DATETIME_SYSTEM_GROUP.WEEK);
		}

		if (type in ATTRIBUTE_SETS.DATE) {
			return createDefaultGroup(DATETIME_SYSTEM_GROUP.MONTH);
		}
	}

	return createDefaultGroup(DEFAULT_SYSTEM_GROUP.OVERLAP);
};

/**
 * Фильтрует виджеты по режиму отображения
 * @param {Array<Widget>} widgets - список виджетов
 * @param {LayoutMode} mode - режим отображения
 * @returns {Array<Widget>}
 */
const getLayoutWidgets = (widgets: Array<Widget>, mode: LayoutMode): Array<Widget> => {
	return widgets.filter(item => (!item.displayMode && mode === LAYOUT_MODE.WEB) || item.displayMode === mode || item.displayMode === DISPLAY_MODE.ANY);
};

/**
 * Сообщает используется ли в наборе данных виджета агрегация в процентах
 * @param {MixedAttribute | null} attribute - атрибут
 * @param {string} aggregation - агрегация атрибута
 * @returns {boolean}
 */
const hasPercent = (attribute: MixedAttribute | null, aggregation: string): boolean => {
	return Boolean(attribute && attribute.type !== ATTRIBUTE_TYPES.COMPUTED_ATTR && aggregation === DEFAULT_AGGREGATION.PERCENT);
};

/**
 * Сообщает об использовании uuid в лейблах
 * @param {Attribute} attribute - атрибут
 * @param {Group} group - группировка
 * @returns {boolean}
 */
const hasUUIDsInLabels = (attribute?: Attribute, group?: Group): boolean => {
	const {dtInterval, metaClass, state} = ATTRIBUTE_TYPES;

	return (attribute && attribute.type in ATTRIBUTE_SETS.REFERENCE)
		|| attribute?.type === metaClass
		|| attribute?.type === state
		|| attribute?.type === dtInterval
		|| group?.way === GROUP_WAYS.CUSTOM;
};

/**
 * Сообщает используется ли в наборе данных виджета агрегация, по которой возвращается интервал в миллисекундах
 * @param {MixedAttribute | null} attribute - атрибут
 * @param {string} aggregation - агрегация атрибута
 * @returns {boolean}
 */
const hasMSInterval = (attribute: MixedAttribute | null, aggregation: string): boolean => {
	return Boolean(attribute && attribute.type === ATTRIBUTE_TYPES.dtInterval && aggregation in INTEGER_AGGREGATION);
};

/**
 * Преобразует интервал из миллисекунд в понятный для пользователя вид
 * @param {number} ms - значение интервала в миллисекундах
 * @returns {string}
 */
const parseMSInterval = (ms: number) => {
	const intervalData = INTERVALS.find(({max, min}) => ms > min && ms < max);

	if (intervalData) {
		const {label, min} = intervalData;
		let intervalValue = ms / min;

		intervalValue = ms > INTERVALS[INTERVALS.length - 1].min ? intervalValue.toFixed(2) : Math.round(intervalValue);

		return `${intervalValue} ${label}`;
	}

	return ms;
};

/**
 * Сообщает используется ли в наборе данных виджета разбивка
 * @param {object} widget - виджет
 * @returns {boolean}
 */
const hasBreakdown = (widget: Object): boolean => widget.id !== NewWidget.id && !!widget.data
		.find(({breakdown, sourceForCompute}) => !sourceForCompute && Array.isArray(breakdown));

/**
 * Проверяет является ли переданный тип агрегации допустимым для выборки топ значений
 * @param {string} aggregation - тип агрегации
 * @returns {boolean}
 */
const isAllowedTopAggregation = (aggregation?: string) => aggregation !== DEFAULT_AGGREGATION.NOT_APPLICABLE;

/**
 * Проверяет принадлежность типа виджета к осевым графикам
 * @param {AnyWidget | string} type - тип виджета
 * @returns {boolean}
 */
const isAxisChart = (type: WidgetType | string): boolean => {
	const {BAR, BAR_STACKED, COLUMN, COLUMN_STACKED, COMBO, LINE} = WIDGET_TYPES;

	return [BAR, BAR_STACKED, COMBO, COLUMN, COLUMN_STACKED, LINE].includes(type);
};

/**
 * Проверяет принадлежность типа виджета к круговым графикам
 * @param {AnyWidget | string} type - тип виджета
 * @returns {boolean}
 */
const isCircleChart = (type: WidgetType | string): boolean => {
	const {DONUT, PIE} = WIDGET_TYPES;

	return [DONUT, PIE].includes(type);
};

/**
 * Проверяет принадлежность типа виджета к горизонтальным графикам
 * @param {AnyWidget} type - тип виджета
 * @returns {boolean}
 */
const isHorizontalChart = (type: WidgetType): boolean => {
	const {BAR, BAR_STACKED} = WIDGET_TYPES;

	return [BAR, BAR_STACKED].includes(type);
};

/**
 * Проверяет принадлежность типа виджета к графикам с накоплением
 * @param {AnyWidget} type - тип виджета
 * @returns {boolean}
 */
const isStackedChart = (type: WidgetType): boolean => {
	const {BAR_STACKED, COLUMN_STACKED} = WIDGET_TYPES;

	return [BAR_STACKED, COLUMN_STACKED].includes(type);
};

/**
 * Проверяет возможность наличия глобальных настроек цветов
 * @param {AnyWidget} type - тип виджета
 * @returns {boolean}
 */
const hasChartColorsSettings = (type: WidgetType): boolean => {
	const {BAR, BAR_STACKED, COLUMN, COLUMN_STACKED} = WIDGET_TYPES;
	const barCharts = [BAR, BAR_STACKED, COLUMN, COLUMN_STACKED];

	return isCircleChart(type) || barCharts.includes(type);
};

export {
	createDefaultGroup,
	getDefaultSystemGroup,
	getLayoutWidgets,
	hasBreakdown,
	hasChartColorsSettings,
	hasUUIDsInLabels,
	hasMSInterval,
	hasPercent,
	isAllowedTopAggregation,
	isAxisChart,
	isCircleChart,
	isHorizontalChart,
	isStackedChart,
	parseMSInterval
};
