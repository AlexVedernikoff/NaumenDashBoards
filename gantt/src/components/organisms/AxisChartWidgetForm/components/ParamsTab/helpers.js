// @flow
import type {ChartSorting} from 'store/widgets/data/types';
import {GROUP_WAYS} from 'store/widgets/constants';
import type {Parameter} from 'store/widgetForms/types';
import {SORTING_VALUES} from 'store/widgets/data/constants';

const getSortValue = (parameters: Array<Parameter>, sorting: ChartSorting) => {
	const {group} = parameters[0];
	const {DEFAULT, INDICATOR} = SORTING_VALUES;
	let {value} = sorting;

	if (group.way === GROUP_WAYS.CUSTOM && sorting.value !== DEFAULT) {
		value = DEFAULT;
	} else if (sorting.value === DEFAULT) {
		value = INDICATOR;
	}

	return value;
};

export {
	getSortValue
};
