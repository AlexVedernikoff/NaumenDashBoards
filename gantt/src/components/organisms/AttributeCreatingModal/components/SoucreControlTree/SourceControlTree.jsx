// @flow
import CreationPanel from 'components/atoms/CreationPanel';
import {getMapValues} from 'helpers';
import type {Props} from './types';
import React, {Fragment, PureComponent} from 'react';
import styles from './styles.less';
import Tree from 'components/molecules/TreeSelect/components/Tree';
import type {Tree as TreeType} from 'components/molecules/TreeSelect/types';

export class SourceControlTree extends PureComponent<Props> {
	countAttributes = (options: TreeType) => getMapValues(options).filter(node => node.parent).length;

	renderAddConstantButton = () => {
		const {onAddConstant, searchValue} = this.props;

		return !searchValue ? <CreationPanel onClick={onAddConstant} text="Добавить константу" /> : null;
	};

	renderFoundInfo = () => {
		const {options, originalOptions, searchValue} = this.props;

		if (searchValue) {
			const countAttributes = this.countAttributes(originalOptions);
			const countFoundAttributes = this.countAttributes(options);
			const info = `Найдено ${countFoundAttributes} из ${countAttributes}`;

			return <div className={styles.searchInfo}>{info}</div>;
		}
	};

	renderSearchInfo = () => {
		const {searchValue} = this.props;

		if (searchValue) {
			return (
				<div className={styles.searchValue}>
					<span>Содержащее: </span>
					{searchValue}
				</div>
			);
		}
	};

	renderTree = () => {
		const {onAddConstant, originalOptions, ...props} = this.props;

		return <Tree {...props} />;
	};

	render () {
		return (
			<Fragment>
				{this.renderSearchInfo()}
				{this.renderTree()}
				{this.renderFoundInfo()}
				{this.renderAddConstantButton()}
			</Fragment>
		);
	}
}

export default SourceControlTree;
