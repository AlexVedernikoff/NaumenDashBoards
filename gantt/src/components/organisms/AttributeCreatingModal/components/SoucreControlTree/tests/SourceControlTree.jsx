import {getProps} from './helpers';
import {getTree} from 'components/molecules/TreeSelect/components/Tree/tests/helpers';
import React from 'react';
import {shallow} from 'enzyme';
import SourceControlTree from 'components/organisms/AttributeCreatingModal/components/SoucreControlTree';

describe('AttributeCreatingModal SourceControlTree', () => {
	it('renders when "searchValue" property is set to empty string', () => {
		const wrapper = shallow(<SourceControlTree {...getProps()} />);

		expect(wrapper).toMatchSnapshot();
	});

	it('renders when "searchValue" property is set to not empty string', () => {
		const wrapper = shallow(<SourceControlTree {...getProps()} searchValue="test" />);

		expect(wrapper).toMatchSnapshot();
	});

	it('method "countAttributes" returns attributes count of the passed options', () => {
		const limit = 4;
		const options = getTree(2, limit);
		const wrapper = shallow(<SourceControlTree {...getProps()} options={options} />);

		expect(wrapper.instance().countAttributes(options)).toEqual(Math.pow(limit, 2));
	});
})

