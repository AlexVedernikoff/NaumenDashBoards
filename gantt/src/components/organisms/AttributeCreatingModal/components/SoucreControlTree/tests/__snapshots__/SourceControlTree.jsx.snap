// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttributeCreatingModal SourceControlTree renders when "searchValue" property is set to empty string 1`] = `
<Fragment>
  <Tree
    components={
      Object {
        "IndicatorsContainer": [Function],
        "LabelContainer": [Function],
        "MenuContainer": [Function],
        "Node": [Function],
        "SearchInput": [Function],
        "Tree": [Function],
        "ValueContainer": [Function],
      }
    }
    getNodeLabel={[Function]}
    getNodeValue={[Function]}
    getOptionLabel={[Function]}
    getOptionValue={[Function]}
    isDisabled={[Function]}
    loading={false}
    multiple={false}
    onSelect={[Function]}
    options={Object {}}
    searchValue=""
    showMore={false}
    value={null}
    values={Array []}
  />
  <CreationPanel
    className=""
    onClick={[MockFunction]}
    text="Добавить константу"
  />
</Fragment>
`;

exports[`AttributeCreatingModal SourceControlTree renders when "searchValue" property is set to not empty string 1`] = `
<Fragment>
  <div
    className="searchValue"
  >
    <span>
      Содержащее: 
    </span>
    test
  </div>
  <Tree
    components={
      Object {
        "IndicatorsContainer": [Function],
        "LabelContainer": [Function],
        "MenuContainer": [Function],
        "Node": [Function],
        "SearchInput": [Function],
        "Tree": [Function],
        "ValueContainer": [Function],
      }
    }
    getNodeLabel={[Function]}
    getNodeValue={[Function]}
    getOptionLabel={[Function]}
    getOptionValue={[Function]}
    isDisabled={[Function]}
    loading={false}
    multiple={false}
    onSelect={[Function]}
    options={Object {}}
    searchValue="test"
    showMore={false}
    value={null}
    values={Array []}
  />
  <div
    className="searchInfo"
  >
    Найдено 0 из 0
  </div>
</Fragment>
`;
